### DISCUSSION TITLE
Manifest.permission.FOREGROUND_SERVICE

### DISCUSSION THREAD

**Discussion by bear101** [BearWare/TeamTalk5#2222] [Q&A]
Hi @poretsky and @beqabeqa473 Build version 126 was reject by Google Play Store because we're using [Manifest.permission.FOREGROUND_SERVICE](https://developer.android.com/reference/android/Manifest.permission#FOREGROUND_SERVICE)

Isn't this permission required when the app goes to the background and the user should still be able to hear voice?

**Comment by poretsky** [comment]
It is not a criminal itself and we declare foreground service type as
required and request specific permission. What is the exact Google
complaint?

**Reply by bear101** [reply]
Issue found: Play Console declaration and/or app description in Google Play Store does not sufficiently demonstrate the use of permission.

We are unable to approve your app’s use of [FOREGROUND_SERVICE](https://notifications.google.com/g/p/ANiao5r-5B7gxTkSaazz5HvaQtBu6hT6l1iyiMqpCGIxDy3NccnzXA7pR21bx3r6Rii__P0xo8QBge7hOlAlOkCj8ExOWLxRqdecZ6yHoY0fpoyUU2IoUUgBScIGXzxtu_WxQ2ZSJkqos1FrQPwGFpwOiaRZ-dr78cxjw-Q69iQy3bTFJy6qRafDdPIIvqv4zq3S_2f7pBIQNNoIAvG3-anC0kkLBiTAsNWwjf5vDi-NnWgl-fECdLUxn1xB7-P8o4bSALh5RTBI6ScUpMwdtATNs0c) permission because we found the following information provided to be insufficient in evaluating compliance with [foreground service permission requirements](https://notifications.google.com/g/p/ANiao5rxiXcq7EjxhvkgsrBS-Kel9AVCySucA6oy_7X_eHChLK2ijU-HwRA1XizMq-MsrkLKWvN0yhzMHAcxblgOfrk7YvDLmMgnHYChvWEzZo_kWGhhlUf-IzPGCEyQDoCAk-YYPL-er_Wysq_Ta47FXpSi9OVjXI7Fv3wM_7W6efnAEE2hHlPGyhxqA7VsToMrKuPtHiybkDOwSdwjES9IV38UbblatTztITbRYlaF59c2NESkrtpLPfEHX6-5qthuOyGH_trshHT8S-R90-9ogsNytOgCkAOKARGZgqcVFjDftNtm):

    The video link you provided in the Play Console declaration and/or
    The app description in the Google Play store.

Issue details

We found an issue in the following area(s):

    Version code 126

To bring your app into compliance, follow these steps:

    Please verify that the version of the app submitted for review is working as intended and that the app description on the Google Play Store accurately reflects the core functionality that requires the use of [foreground service permission](https://notifications.google.com/g/p/ANiao5qkXbrKzMemRTcEJatwArjqZH5m1Yv1sELidQ0S6sIY8NRHWhvMEMI5I5r7Hov5KprgefWxd_iqubn9m8VhNqmN0FwDiKA8jX-VvRreW4swQZZYB6GtAvM_Q4xrpEwfqNW5V2nQ8luxm3OJ9oF3uY3YoTFsAt8-sUomum5W_sNz024--JiQjHRMMK3g-crc8CuJEKfh-pRj0xI3tH1Aqh4-1c7Q3L4Yl1v5Y_QDSBnWipUeu5i8fXp9ZVwaaGCCArRrSQs).
    Submit an [updated declaration](https://notifications.google.com/g/p/ANiao5p7ISBBngH32LMmXMC_5qk-kNmhMsLO5mZgkbtjz3M5K6RcIC6rhtsZbXjp7suDCtAILUOjrxCPrNNi4YZAkotmbdmXkKMzkSP_mUTwXnMpkjcA1QBFRxQSGw_A-Z0KZJSvrHI6oODYGJxVPaq9XSmynT_JAUNjFAN9do7WHTKABsD0icZcSJOniE7-8_CjPT11l-DEs9xgyoc7l1o59t319jcBLWNdRwwY4M_WO7U3oohjtgEHPoCN7nrt0aLGVulvMEEh37Au) with clear and relevant video demonstrating the core feature’s dependency on the [foreground service permission](https://notifications.google.com/g/p/ANiao5qkXbrKzMemRTcEJatwArjqZH5m1Yv1sELidQ0S6sIY8NRHWhvMEMI5I5r7Hov5KprgefWxd_iqubn9m8VhNqmN0FwDiKA8jX-VvRreW4swQZZYB6GtAvM_Q4xrpEwfqNW5V2nQ8luxm3OJ9oF3uY3YoTFsAt8-sUomum5W_sNz024--JiQjHRMMK3g-crc8CuJEKfh-pRj0xI3tH1Aqh4-1c7Q3L4Yl1v5Y_QDSBnWipUeu5i8fXp9ZVwaaGCCArRrSQs).
    Make any necessary corrections or modifications to meet the policy requirement.
    Ensure [valid testing credentials](https://notifications.google.com/g/p/ANiao5o8SZMNLdrLshuVR94penZssxBdczs0LF8KIPhUX6E6lFbXmc4oNwBgbXOlm2evR4VbeP0z13LPkqu0l8JX4TyF9BtOwJvBvoheCJuGyTmcf3Sj6Mze4l0kMJXLie4tvYc4YlTTM3TAsQzd1vZV6vXFWmkS3z3Q3pXj2KXFqwUqrbRiapHyMr3ijcrhKD4aKgDgjSjv8RY3tSno2V2cYkvmtvJoVhRcFjhSHourB_OSo-IWnxF_Lt6gpveuubeD0sRX4ZJYnAJR) are provided.

About the Permissions for Foreground Services

The Foreground Service permission ensures the appropriate use of user-facing foreground services. For apps targeting Android 14 and above, you must specify a valid foreground service type for each foreground service used in your app, and declare the foreground service permission that is appropriate for that type. For example, if your app’s use case requires map geolocation, you must declare the FOREGROUND_SERVICE_LOCATION permission in your app’s manifest.
With the exceptions of systemExempted and shortService foreground service types, apps are only allowed to declare a foreground service permission if the use:

    Provides a feature that is beneficial to the user and relevant to the core functionality of the app.
    Is initiated by the user or is user perceptible (for example, audio from playing a song, cast media to another device, accurate and clear user notification, user request to upload a photo to the cloud).
    Can be terminated or stopped by the user.
    Can’t be interrupted or deferred by the system without causing a negative user experience or causing the user anticipated feature to not work as intended (for example, a phone call needs to start immediately and can’t be deferred by the system).
    Runs only for as long as necessary to complete the task.

The use of foreground service is further explained on this [Play Console Help Center](https://notifications.google.com/g/p/ANiao5rxiXcq7EjxhvkgsrBS-Kel9AVCySucA6oy_7X_eHChLK2ijU-HwRA1XizMq-MsrkLKWvN0yhzMHAcxblgOfrk7YvDLmMgnHYChvWEzZo_kWGhhlUf-IzPGCEyQDoCAk-YYPL-er_Wysq_Ta47FXpSi9OVjXI7Fv3wM_7W6efnAEE2hHlPGyhxqA7VsToMrKuPtHiybkDOwSdwjES9IV38UbblatTztITbRYlaF59c2NESkrtpLPfEHX6-5qthuOyGH_trshHT8S-R90-9ogsNytOgCkAOKARGZgqcVFjDftNtm) page.

**Comment by poretsky** [comment]
I still don't understand what else is incorrect. It seems that a fix is
suggested in the section "To bring your app into compliance, follow
these steps", but the links there are unavailable without Google play
account.

**Reply by bear101** [reply]
What is it that TeamTalkService.displayNotification() does? If I comment out the code to startForeground() then I don't see a difference.

**Comment by poretsky** [comment]
Being passed true as an argument TeamTalkService.displayNotification()
displays or updates (if it is displayed already) TeamTalk notification
in the notification panel. By the way, initial notification display is
accompanied by turning the foreground mode on. On the contrary, being
passed false it removes TeamTalk notification from the notification
panel and turns foreground mode off.

**Reply by bear101** [reply]
If I should make a video of how displayNotification() is used then how should I do it? I don't see any notifications on the Pixel 5 API 29 or Pixel 7 Pro API 34.

**Reply by bear101** [reply]
If I comment out the call to startForeground(UI_WIDGET_ID, widget.build(), ServiceInfo.FOREGROUND_SERVICE_TYPE_MANIFEST) doesn't it still work as it should? I don't see a difference.

**Comment by poretsky** [comment]
It should not cause a visible runtime difference, but starting
foreground without specifying service type (third argument) is somewhat
deprecated in modern Android. Thus, I hope that this variant will not be
discarded by Google play.

**Reply by bear101** [reply]
Isn't it better to not require the permission then? We should not ask for permissions that are not needed.

**Comment by poretsky** [comment]
Of course, but I still think that the service should be foreground and,
thus, the corresponding permissions are necessary.

**Comment by poretsky** [comment]
As I understand, this could happen when posting notifications is
disabled for the application. In that case the service itself goes
foreground, but without showing the notification, being visible only in
the task manager.

I found a good detailed guide about foreground services in Android at

***@***.***/guide-to-foreground-services-on-android-9d0127dc8f9ar

As I conclude from it, the problem is not in the permission request
itself, but in the fact that some additional information should be
provided.

Though I recommend merging #2223 anyway, since it increases SDK upgrade
tolerance at least.
